### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by beteland123**
**Note :**  Dear Reviewer I am writing to request a code review for a project I have been working on. Unfortunately, I have not been assigned a partner for pair programming, and I have been working on the project independently. This is my collaboration information

colb_info

In this branch I -
- Create integration tests for all views, Using Capybara to write integration tests for each view in the project.
 for User index page, user show page, User post index page and Post show page
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user
- before fix 

bef_post_index

bef_post_show

- after fix

aft_post_index

aft_post_Show

### CODE REVIEW

**Review by Qoosim — COMMENTED** [review]
#  *Status: Invalid Request* :red_circle: 

Hi @beteland123, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.



## Invalid Code Review Request

You have submitted a project, that is pair-programming but contain contributions from only one partner. Kindly provide screenshot evident from student success that states that you should continue working solo.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/views/user/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it 'it show the profile picture of the user' do
+    expect(page).to have_css("img[src*='menen.jpg']")
+    expect(page).to have_css("img[src*='gelila.jpg']")
+  end
+
+  it 'provides links to user show pages' do
+    expect(page).to have_link(user1.name, href: user_path(user1))
+    expect(page).to have_link(user2.name, href: user_path(user2))
```
Comment (mckent05) [inline spec/views/user/user_index_spec.rb]: - Good job done so far, but according to the project requirements, please test that the page redirects to the user's show page when you click on a user. You can use the `have_current_path` matcher to test for the path. Please fix this 🙏

File: spec/views/user/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  it 'show the bio of the user' do
+    expect(page).to have_content(user.bio)
+  end
+
+  it 'shows the user first three posts' do
+    user.posts.limit(3).each do |post|
+      expect(page).to have_content(post.title)
+    end
+  end
```
Comment (mckent05) [inline spec/views/user/user_show_spec.rb]: - You have done well so far, remember you have a method, `recent_posts`, in your user model that gives the first three posts for a user. I suggest you test this method to check for a user's first three posts. Please fix this 🙏

Comment (mckent05) [inline spec/views/user/user_show_spec.rb]: - You have done well so far, according to the project requirements, you are missing some required tests. Please fix this 🙏. The missing tests are:
* Tests that assert you can see a button that lets me view all of a user's posts.
* Tests that redirect to the post's show page when you click on a user's post.
* Tests that redirect to the user's post's index page when you click on a **see all posts**

File: spec/views/post/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    it 'I can see a post title.' do
+      expect(page).to have_content(post.title)
+    end
+    it 'displays some of the post\'s body' do
+      expect(page).to have_selector('p', text: 'Post Text')
+    end
+    it 'I can see the first comments on a post.' do
+      post.comments.first(5).each do |comment|
+        expect(page).to have_content(comment.text)
+      end
```
Comment (mckent05) [inline spec/views/post/post_index_spec.rb]: - You have done well so far, but I suggest that you use the method, `recent_comments` in the posts model that returns the first 5 comments for a post. Please fix this 😄

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @beteland123, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job fixing the N+1 query problem and adding integration tests to your app 👏 👏 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/post/post_index_spec.rb (hunk 3 above)
Comment (beteland123) [reply spec/views/post/post_index_spec.rb]: Hello @mckent05, Thank you for the review it was really helpful 💯 
I already implemented all the required changes.

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @beteland123 , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Great job! All requested changes implemented 👍 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**